File: src/main/java/org/tobi29/scapes/client/gui/GuiPlugins.java
/*
 * Copyright 2012-2015 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.tobi29.scapes.client.gui;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.tobi29.scapes.engine.GameState;
import org.tobi29.scapes.engine.gui.*;
import org.tobi29.scapes.engine.opengl.texture.*;
import org.tobi29.scapes.engine.utils.Pair;
import org.tobi29.scapes.plugins.PluginFile;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.zip.ZipFile;

public class GuiPlugins extends GuiMenu {
    private static final Logger LOGGER =
            LoggerFactory.getLogger(GuiPlugins.class);
    private final Path path;
    private final GuiComponentScrollPaneViewport scrollPane;

    @SuppressWarnings("unchecked")
    public GuiPlugins(GameState state, Gui previous, GuiStyle style) {
        super(state, "Plugins", previous, style);
        path = state.engine().home().resolve("plugins");
        scrollPane = pane.addVert(16, 5,
                p -> new GuiComponentScrollPane(p, 368, 290, 70)).viewport();
        GuiComponentTextButton add = pane.addVert(112, 5,
                p -> new GuiComponentTextButton(p, 176, 30, 18, "Add"));
        add.addLeftClick(event -> {
            try {
                Path[] imports = state.engine().container().openFileDialog(
                        new Pair[]{new Pair<>("*.jar", "Jar Archive")},
                        "Import plugin", true);
                for (Path copy : imports) {
                    Files.copy(copy, path.resolve(copy.getFileName()));
                }
                updatePlugins();
            } catch (IOException e) {
                LOGGER.warn("Failed to import plugin: {}", e.toString());
            }
        });
        updatePlugins();
    }

    private void updatePlugins() {
        try {
            scrollPane.removeAll();
            for (Path file : Files.newDirectoryStream(path)) {
                if (Files.isRegularFile(file) && !Files.isHidden(file)) {
                    scrollPane.addVert(0, 0, p -> new Element(p, file));
                }
            }
        } catch (IOException e) {
            LOGGER.warn("Failed to read plugins: {}", e.toString());
        }
    }

    private class Element extends GuiComponentPane {
        private Texture texture;

        public Element(GuiLayoutData parent, Path path) {
            super(parent, 378, 70);
            try (ZipFile zip = new ZipFile(path.toFile())) {
                texture = new TextureFile(
                        zip.getInputStream(zip.getEntry("Icon.png")), 0,
                        TextureFilter.LINEAR, TextureFilter.LINEAR,
                        TextureWrap.CLAMP, TextureWrap.CLAMP);
                PluginFile plugin = new PluginFile(path);
                add(70, 20, p -> new GuiComponentTextButton(p, 180, 30, 18,
                        plugin.name()));
            } catch (IOException e) {
                texture = new TextureCustom(1, 1);
            }
            GuiComponentTextButton delete = add(260, 20,
                    p -> new GuiComponentTextButton(p, 80, 30, 18, "Delete"));
            delete.addLeftClick(event -> {
                try {
                    Files.delete(path);
                    scrollPane.remove(this);
                } catch (IOException e) {
                    LOGGER.warn("Failed to delete plugin: {}", e.toString());
                }
            });
            add(15, 15, p -> new GuiComponentIcon(p, 40, 40, texture));
        }
    }
}

File: ScapesCore/src/main/kotlin/org/tobi29/scapes/client/gui/GuiPlugins.kt
/*
 * Copyright 2012-2017 Tobi29
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.tobi29.scapes.client.gui

import kotlinx.coroutines.experimental.launch
import kotlinx.coroutines.experimental.runBlocking
import org.tobi29.scapes.client.DialogProvider
import org.tobi29.scapes.client.ScapesClient
import org.tobi29.scapes.engine.GameState
import org.tobi29.scapes.engine.graphics.TextureFilter
import org.tobi29.scapes.engine.graphics.TextureWrap
import org.tobi29.scapes.engine.gui.*
import org.tobi29.scapes.engine.resource.Resource
import org.tobi29.scapes.engine.utils.graphics.decodePNG
import org.tobi29.scapes.engine.utils.io.*
import org.tobi29.scapes.engine.utils.io.filesystem.*
import org.tobi29.scapes.engine.utils.logging.KLogging
import org.tobi29.scapes.plugins.PluginFile
import org.tobi29.scapes.plugins.Plugins
import java.util.zip.ZipFile

class GuiPlugins(state: GameState,
                 previous: Gui,
                 style: GuiStyle) : GuiMenuDouble(
        state, "Plugins", "Add", "Back", previous, style) {
    private val path: FilePath
    private val scrollPane: GuiComponentScrollPaneViewport

    init {
        val scapes = engine[ScapesClient.COMPONENT]
        path = scapes.home.resolve("plugins")
        scrollPane = pane.addVert(16.0, 5.0, -1.0, -1.0) {
            GuiComponentScrollPane(it, 70)
        }.viewport
        updatePlugins()

        save.on(GuiEvent.CLICK_LEFT) {
            try {
                state.engine[DialogProvider.COMPONENT].openPluginDialog { _, stream ->
                    launch(engine.taskExecutor) {
                        val temp = createTempFile("Plugin", ".jar")
                        write(temp) { output ->
                            stream.process { output.put(it) }
                        }
                        val newPlugin = PluginFile.loadFile(temp)
                        move(temp, path.resolve(newPlugin.id() + ".jar"))
                        updatePlugins()
                    }
                }
            } catch (e: IOException) {
                logger.warn { "Failed to import plugin: $e" }
            }
        }
    }

    private fun updatePlugins() {
        launch(engine.taskExecutor) {
            val files = Plugins.installed(
                    path).asSequence().sortedBy { it.name() }
            synchronized(this@GuiPlugins) {
                scrollPane.removeAll()
                files.forEach { file ->
                    scrollPane.addVert(0.0, 0.0, -1.0, 70.0) {
                        Element(it, file)
                    }
                }
            }
        }
    }

    private inner class Element(parent: GuiLayoutData,
                                plugin: PluginFile) : GuiComponentGroupSlab(
            parent) {
        init {
            val icon = addHori(15.0, 15.0, 40.0, -1.0) { GuiComponentIcon(it) }
            val label = addHori(5.0, 20.0, -1.0, -1.0) {
                button(it, "Invalid plugin")
            }
            val delete = addHori(5.0, 20.0, 80.0, -1.0) {
                button(it, "Delete")
            }

            if (plugin.file() != null) {
                delete.on(GuiEvent.CLICK_LEFT) {
                    try {
                        plugin.file()?.let { delete(it) }
                        scrollPane.remove(this)
                    } catch (e: IOException) {
                        logger.warn { "Failed to delete plugin: $e" }
                    }
                }
                try {
                    label.setText(plugin.name())
                    plugin.file()?.let {
                        ZipFile(it.toFile()).use { zip ->
                            val stream = BufferedReadChannelStream(
                                    Channels.newChannel(zip.getInputStream(
                                            zip.getEntry(
                                                    "scapes/plugin/Icon.png"))).toChannel())
                            val image = runBlocking { decodePNG(stream) }
                            val texture = state.engine.graphics.createTexture(
                                    image, 0, TextureFilter.LINEAR,
                                    TextureFilter.LINEAR, TextureWrap.CLAMP,
                                    TextureWrap.CLAMP)
                            icon.texture = Resource(texture)
                        }
                    }
                } catch (e: IOException) {
                }
            }
        }
    }

    companion object : KLogging()
}
